File: backend/src/main/java/com/icms/model/Complaint.java
package com.icms.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "complaints")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Complaint {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String complaintId;

    @ManyToOne
    private User user;

    private String category;
    private String location;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    private Priority priority;

    @Enumerated(EnumType.STRING)
    private Status status;

    private String contactNumber;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    // Transient field for frontend display
    @Transient
    private String studentName;

    public enum Priority { LOW, MEDIUM, HIGH, URGENT }

    public enum Status { PENDING, IN_PROGRESS, RESOLVED }

    @PrePersist
    void onCreate() {
        createdAt = LocalDateTime.now();
    }

    @PostLoad
    void postLoad() {
        if (user != null) {
            this.studentName = user.getName();
        }
    }
}

File: backend/models/Complaint.js
const mongoose = require('mongoose');

const complaintSchema = new mongoose.Schema({
    complaint_id: {
        type: String,
        required: true,
        unique: true,
        maxlength: 20
    },
    user_id: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    student_name: {
        type: String,
        required: true,
        maxlength: 100
    },
    category: {
        type: String,
        required: true,
        maxlength: 50
    },
    location: {
        type: String,
        required: true,
        maxlength: 255
    },
    description: {
        type: String,
        required: true
    },
    priority: {
        type: String,
        enum: ['LOW', 'MEDIUM', 'HIGH', 'URGENT'],
        default: 'MEDIUM',
        required: true
    },
    status: {
        type: String,
        enum: ['PENDING', 'IN_PROGRESS', 'RESOLVED'],
        default: 'PENDING',
        required: true
    },
    contact_number: {
        type: String,
        maxlength: 20
    },
    imagePath: {
        type: String,
        default: null
    }
}, {
    timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' }
});

const Complaint = mongoose.model('Complaint', complaintSchema);
module.exports = Complaint;
